xdsl/dialects/builtin.py

```diff
@@ -228,20 +226,13 @@ def infer(self, context: ConstraintContext) -> ArrayAttr[AttributeCovT]:
     def get_unique_base(self) -> type[Attribute] | None:
         return ArrayAttr
 
-    @dataclass(frozen=True)
-    class _Extractor(VarExtractor[Attribute]):
-        inner: VarExtractor[Sequence[Attribute]]
+    def variables(self) -> set[str]:
+        return self.elem_range_constraint.variables()
```

Comment:
how do you feel about this being an iterator instead? the caller can then call `set(constr.variables)` if they feel like it, and we avoid building and throwing away a ton of objects

Reply:
Would likely be fine as an iterator, sure

Reply:
Would an iterator really be better here? What would we gain?
I'm not sure that a `yield` is really faster than creating a small `set`?

Reply:
Yeah I started coding it up and you're just replacing set generation with python generator generation everywhere.

Reply:
I imagine if you really care about speed then the ideal is passing in a set which gets mutated, but that feels less nice to me.

Reply:
I just think it's cleaner, the implementation should be simpler, and it's more direct

Reply:
just to be clear, you'd prefer:
```python
yield from a
yield from b
```
to 
```python
return a | b
```

Reply:
Wait, what would happen when there are twice the same variable? The iterator way wouldn't work no?

Reply:
I think the idea is that it would duplicate the variable, and you generate one set from the iterator at the end which would remove the duplicate

Reply:
I think this is quite yuck in my opinion. This is quite error-prone no?

Reply:
I think it's the opposite, it lets you do things like detect that there are duplicate values, if you feel like it.

Reply:
I'm not sure I see the utility in that.

Reply:
I mean I also am not sure but it feels like the least opinionated way of implementing this function, and should also be more efficient and easy to read